File: src/client/types/media.ts
import type { ParseMode } from "../constants.js";
import type { MessageEntity } from "./messages.js";
import type { InputFile } from "../../utils/http.js";

export type { InputFile };

/**
 * Represents one size of a photo or a file / sticker thumbnail.
 */
export interface PhotoSize {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Photo width. */
  width: number;
  /** Photo height. */
  height: number;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents an audio file to be treated as music by the Telegram clients.
 */
export interface Audio {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Duration of the audio in seconds as defined by sender. */
  duration: number;
  /** Performer of the audio as defined by sender or by audio tags. */
  performer?: string;
  /** Title of the audio as defined by sender or by audio tags. */
  title?: string;
  /** Original filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
  /** Thumbnail of the album cover to which the music file belongs. */
  thumbnail?: PhotoSize;
}

/**
 * Represents a general file (as opposed for photos or audio files).
 */
export interface Document {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Document thumbnail as defined by sender. */
  thumbnail?: PhotoSize;
  /** Original filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a video file.
 */
export interface Video {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width as defined by sender. */
  width: number;
  /** Video height as defined by sender. */
  height: number;
  /** Duration of the video in seconds as defined by sender. */
  duration: number;
  /** Video thumbnail. */
  thumbnail?: PhotoSize;
  /** Original filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents an animation file (GIF or H.264/MPEG-4 AVC video without sound).
 */
export interface Animation {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width as defined by sender. */
  width: number;
  /** Video height as defined by sender. */
  height: number;
  /** Duration of the video in seconds as defined by sender. */
  duration: number;
  /** Animation thumbnail. */
  thumbnail?: PhotoSize;
  /** Original animation filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a voice note.
 */
export interface Voice {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Duration of the audio in seconds as defined by sender. */
  duration: number;
  /** MIME type of the audio as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a video message (round video).
 */
export interface VideoNote {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width and height (diameter of the video message) as defined by sender. */
  length: number;
  /** Duration of the video in seconds as defined by sender. */
  duration: number;
  /** Video thumbnail. */
  thumbnail?: PhotoSize;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a Live Photo message object (Bot API 10.3+).
 */
export interface LivePhoto {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file. */
  file_unique_id: string;
  /** Photo width. */
  width: number;
  /** Photo height. */
  height: number;
  /** Available sizes of the photo. */
  photo: PhotoSize[];
  /** Video file associated with the live photo. */
  video: Video;
}

/**
 * Represents a photo to be sent as part of a media group or edited media.
 */
export interface InputMediaPhoto {
  /** Type of the result, must be 'photo'. */
  type: "photo";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile | unknown;
  /** Caption of the photo to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the photo caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Pass True, if the caption must be shown above the message media. */
  show_caption_above_media?: boolean;
  /** Pass True, if the photo needs to be covered with a spoiler animation. */
  has_spoiler?: boolean;
}

/**
 * Represents a video to be sent as part of a media group or edited media.
 */
export interface InputMediaVideo {
  /** Type of the result, must be 'video'. */
  type: "video";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile | unknown;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile | unknown;
  /** Caption of the video to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the video caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Pass True, if the caption must be shown above the message media. */
  show_caption_above_media?: boolean;
  /** Video width. */
  width?: number;
  /** Video height. */
  height?: number;
  /** Video duration in seconds. */
  duration?: number;
  /** Pass True, if the uploaded video is suitable for streaming. */
  supports_streaming?: boolean;
  /** Pass True, if the video needs to be covered with a spoiler animation. */
  has_spoiler?: boolean;
}

/**
 * Represents an animation file to be sent as part of a media group or edited media.
 */
export interface InputMediaAnimation {
  /** Type of the result, must be 'animation'. */
  type: "animation";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile | unknown;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile | unknown;
  /** Caption of the animation to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the animation caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Pass True, if the caption must be shown above the message media. */
  show_caption_above_media?: boolean;
  /** Animation width. */
  width?: number;
  /** Animation height. */
  height?: number;
  /** Animation duration in seconds. */
  duration?: number;
  /** Pass True, if the animation needs to be covered with a spoiler animation. */
  has_spoiler?: boolean;
}

/**
 * Represents an audio file to be treated as music to be sent as part of a media group or edited media.
 */
export interface InputMediaAudio {
  /** Type of the result, must be 'audio'. */
  type: "audio";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile | unknown;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile | unknown;
  /** Caption of the audio to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the audio caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Duration of the audio in seconds. */
  duration?: number;
  /** Performer of the audio. */
  performer?: string;
  /** Title of the audio. */
  title?: string;
}

/**
 * Represents a general file to be sent as part of a media group or edited media.
 */
export interface InputMediaDocument {
  /** Type of the result, must be 'document'. */
  type: "document";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile | unknown;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile | unknown;
  /** Caption of the document to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the document caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Disables automatic server-side content type detection for files. */
  disable_content_type_detection?: boolean;
}

/**
 * Union of all input media types that can be sent in a media group or editMessageMedia.
 */
export type InputMedia =
  InputMediaPhoto | InputMediaVideo | InputMediaAnimation | InputMediaAudio | InputMediaDocument;

File: packages/python/src/telebot_py/bot/media.py
"""Media-sending Bot API methods (parity with packages/go/pkg/bot/media.go)."""

from __future__ import annotations

from collections.abc import Sequence

from telebot_py.bot.base import (
    MarkupLike,
    Requester,
    clean_payload,
    parse_result,
    to_wire,
)
from telebot_py.types.common import MessageEntity
from telebot_py.types.message import Message
from telebot_py.types.message_extras import EphemeralMessageParameters, ReplyParameters
from telebot_py.types.suggested_post_types import SuggestedPostParameters


class MediaMixin(Requester):
    """Bot methods for sending audio, video, animations, voice, and video notes.

    Media parameters accept ``file_id`` strings or HTTP URLs; multipart file
    uploads are intentionally out of scope (JSON payloads only).
    """

    async def send_audio(
        self,
        chat_id: int | str,
        audio: str,
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        direct_messages_topic_id: int | None = None,
        ephemeral_message_parameters: EphemeralMessageParameters | MarkupLike | None = None,
        caption: str | None = None,
        parse_mode: str | None = None,
        caption_entities: Sequence[MessageEntity] | None = None,
        duration: int | None = None,
        performer: str | None = None,
        title: str | None = None,
        thumbnail: str | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
        allow_paid_broadcast: bool | None = None,
        message_effect_id: str | None = None,
        suggested_post_parameters: SuggestedPostParameters | MarkupLike | None = None,
        reply_parameters: ReplyParameters | MarkupLike | None = None,
        reply_markup: MarkupLike | None = None,
    ) -> Message:
        """Send an audio file by ``file_id`` or HTTP URL.

        Example:
            >>> msg = await bot.send_audio(123456, "audio_file_id", caption="song")

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            audio: Audio file to send as a ``file_id`` string or HTTP URL.
            business_connection_id: Unique identifier of the business
                connection on behalf of which the message will be sent.
            message_thread_id: Unique identifier for the target message thread.
            direct_messages_topic_id: Identifier of the direct messages topic to
                which the message will be sent; required if the message is sent
                to a direct messages chat.
            ephemeral_message_parameters: EphemeralMessageParameters as a
                ``to_dict`` object or dict.
            caption: Audio caption, 0-1024 characters.
            parse_mode: Parse mode for the caption.
            caption_entities: Special entities for the caption.
            duration: Duration of the audio in seconds.
            performer: Performer of the audio.
            title: Track name.
            thumbnail: Thumbnail file_id or URL of the file sent.
            disable_notification: Send silently.
            protect_content: Protect the content from forwarding and saving.
            allow_paid_broadcast: Pass True to ignore broadcasting limits for a
                fee of 0.1 Telegram Stars per message.
            message_effect_id: Unique identifier of the message effect to add.
            suggested_post_parameters: SuggestedPostParameters as a ``to_dict``
                object or dict; for direct messages chats only.
            reply_parameters: Description of the message to reply to, as a
                ``ReplyParameters`` object or a mapping.
            reply_markup: Markup for the message; dict or ``to_dict`` object.

        Returns:
            The sent Message.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#sendaudio
        """
        payload = clean_payload(
            chat_id=chat_id,
            audio=audio,
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            direct_messages_topic_id=direct_messages_topic_id,
            ephemeral_message_parameters=to_wire(ephemeral_message_parameters),
            caption=caption,
            parse_mode=parse_mode,
            caption_entities=[entity.to_dict() for entity in caption_entities]
            if caption_entities is not None
            else None,
            duration=duration,
            performer=performer,
            title=title,
            thumbnail=thumbnail,
            disable_notification=disable_notification,
            protect_content=protect_content,
            allow_paid_broadcast=allow_paid_broadcast,
            message_effect_id=message_effect_id,
            suggested_post_parameters=to_wire(suggested_post_parameters),
            reply_parameters=to_wire(reply_parameters),
            reply_markup=to_wire(reply_markup),
        )
        return parse_result(Message, await self.request("sendAudio", payload))

    async def send_video(
        self,
        chat_id: int | str,
        video: str,
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        direct_messages_topic_id: int | None = None,
        ephemeral_message_parameters: EphemeralMessageParameters | MarkupLike | None = None,
        duration: int | None = None,
        width: int | None = None,
        height: int | None = None,
        thumbnail: str | None = None,
        caption: str | None = None,
        parse_mode: str | None = None,
        caption_entities: Sequence[MessageEntity] | None = None,
        show_caption_above_media: bool | None = None,
        has_spoiler: bool | None = None,
        supports_streaming: bool | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
        allow_paid_broadcast: bool | None = None,
        message_effect_id: str | None = None,
        suggested_post_parameters: SuggestedPostParameters | MarkupLike | None = None,
        reply_parameters: ReplyParameters | MarkupLike | None = None,
        reply_markup: MarkupLike | None = None,
    ) -> Message:
        """Send a video file by ``file_id`` or HTTP URL.

        Example:
            >>> msg = await bot.send_video(123456, "video_file_id", width=1920, height=1080)

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            video: Video to send as a ``file_id`` string or HTTP URL.
            business_connection_id: Unique identifier of the business
                connection on behalf of which the message will be sent.
            message_thread_id: Unique identifier for the target message thread.
            direct_messages_topic_id: Identifier of the direct messages topic to
                which the message will be sent; required if the message is sent
                to a direct messages chat.
            ephemeral_message_parameters: EphemeralMessageParameters as a
                ``to_dict`` object or dict.
            duration: Duration of the video in seconds.
            width: Video width.
            height: Video height.
            thumbnail: Thumbnail file_id or URL of the file sent.
            caption: Video caption, 0-1024 characters.
            parse_mode: Parse mode for the caption.
            caption_entities: Special entities for the caption.
            show_caption_above_media: Show the caption above the media.
            has_spoiler: Mark the video as covered with a spoiler.
            supports_streaming: Pass True if the video is suitable for
                streaming.
            disable_notification: Send silently.
            protect_content: Protect the content from forwarding and saving.
            allow_paid_broadcast: Pass True to ignore broadcasting limits for a
                fee of 0.1 Telegram Stars per message.
            message_effect_id: Unique identifier of the message effect to add.
            suggested_post_parameters: SuggestedPostParameters as a ``to_dict``
                object or dict; for direct messages chats only.
            reply_parameters: Description of the message to reply to, as a
                ``ReplyParameters`` object or a mapping.
            reply_markup: Markup for the message; dict or ``to_dict`` object.

        Returns:
            The sent Message.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#sendvideo
        """
        payload = clean_payload(
            chat_id=chat_id,
            video=video,
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            direct_messages_topic_id=direct_messages_topic_id,
            ephemeral_message_parameters=to_wire(ephemeral_message_parameters),
            duration=duration,
            width=width,
            height=height,
            thumbnail=thumbnail,
            caption=caption,
            parse_mode=parse_mode,
            caption_entities=[entity.to_dict() for entity in caption_entities]
            if caption_entities is not None
            else None,
            show_caption_above_media=show_caption_above_media,
            has_spoiler=has_spoiler,
            supports_streaming=supports_streaming,
            disable_notification=disable_notification,
            protect_content=protect_content,
            allow_paid_broadcast=allow_paid_broadcast,
            message_effect_id=message_effect_id,
            suggested_post_parameters=to_wire(suggested_post_parameters),
            reply_parameters=to_wire(reply_parameters),
            reply_markup=to_wire(reply_markup),
        )
        return parse_result(Message, await self.request("sendVideo", payload))

    async def send_animation(
        self,
        chat_id: int | str,
        animation: str,
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        direct_messages_topic_id: int | None = None,
        ephemeral_message_parameters: EphemeralMessageParameters | MarkupLike | None = None,
        duration: int | None = None,
        width: int | None = None,
        height: int | None = None,
        thumbnail: str | None = None,
        caption: str | None = None,
        parse_mode: str | None = None,
        caption_entities: Sequence[MessageEntity] | None = None,
        show_caption_above_media: bool | None = None,
        has_spoiler: bool | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
        allow_paid_broadcast: bool | None = None,
        message_effect_id: str | None = None,
        suggested_post_parameters: SuggestedPostParameters | MarkupLike | None = None,
        reply_parameters: ReplyParameters | MarkupLike | None = None,
        reply_markup: MarkupLike | None = None,
    ) -> Message:
        """Send an animation (GIF or H.264/MPEG-4 AVC without sound).

        Example:
            >>> msg = await bot.send_animation(123456, "anim_file_id", has_spoiler=True)

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            animation: Animation to send as a ``file_id`` string or HTTP URL.
            business_connection_id: Unique identifier of the business
                connection on behalf of which the message will be sent.
            message_thread_id: Unique identifier for the target message thread.
            direct_messages_topic_id: Identifier of the direct messages topic to
                which the message will be sent; required if the message is sent
                to a direct messages chat.
            ephemeral_message_parameters: EphemeralMessageParameters as a
                ``to_dict`` object or dict.
            duration: Duration of the animation in seconds.
            width: Animation width.
            height: Animation height.
            thumbnail: Thumbnail file_id or URL of the file sent.
            caption: Animation caption, 0-1024 characters.
            parse_mode: Parse mode for the caption.
            caption_entities: Special entities for the caption.
            show_caption_above_media: Show the caption above the media.
            has_spoiler: Mark the animation as covered with a spoiler.
            disable_notification: Send silently.
            protect_content: Protect the content from forwarding and saving.
            allow_paid_broadcast: Pass True to ignore broadcasting limits for a
                fee of 0.1 Telegram Stars per message.
            message_effect_id: Unique identifier of the message effect to add.
            suggested_post_parameters: SuggestedPostParameters as a ``to_dict``
                object or dict; for direct messages chats only.
            reply_parameters: Description of the message to reply to, as a
                ``ReplyParameters`` object or a mapping.
            reply_markup: Markup for the message; dict or ``to_dict`` object.

        Returns:
            The sent Message.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#sendanimation
        """
        payload = clean_payload(
            chat_id=chat_id,
            animation=animation,
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            direct_messages_topic_id=direct_messages_topic_id,
            ephemeral_message_parameters=to_wire(ephemeral_message_parameters),
            duration=duration,
            width=width,
            height=height,
            thumbnail=thumbnail,
            caption=caption,
            parse_mode=parse_mode,
            caption_entities=[entity.to_dict() for entity in caption_entities]
            if caption_entities is not None
            else None,
            show_caption_above_media=show_caption_above_media,
            has_spoiler=has_spoiler,
            disable_notification=disable_notification,
            protect_content=protect_content,
            allow_paid_broadcast=allow_paid_broadcast,
            message_effect_id=message_effect_id,
            suggested_post_parameters=to_wire(suggested_post_parameters),
            reply_parameters=to_wire(reply_parameters),
            reply_markup=to_wire(reply_markup),
        )
        return parse_result(Message, await self.request("sendAnimation", payload))

    async def send_voice(
        self,
        chat_id: int | str,
        voice: str,
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        direct_messages_topic_id: int | None = None,
        ephemeral_message_parameters: EphemeralMessageParameters | MarkupLike | None = None,
        caption: str | None = None,
        parse_mode: str | None = None,
        caption_entities: Sequence[MessageEntity] | None = None,
        duration: int | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
        allow_paid_broadcast: bool | None = None,
        message_effect_id: str | None = None,
        suggested_post_parameters: SuggestedPostParameters | MarkupLike | None = None,
        reply_parameters: ReplyParameters | MarkupLike | None = None,
        reply_markup: MarkupLike | None = None,
    ) -> Message:
        """Send an audio file for display as a voice message.

        Example:
            >>> msg = await bot.send_voice(123456, "voice_file_id", duration=10)

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            voice: Voice message to send as a ``file_id`` string or HTTP URL;
                OGG encoded with OPUS.
            business_connection_id: Unique identifier of the business
                connection on behalf of which the message will be sent.
            message_thread_id: Unique identifier for the target message thread.
            direct_messages_topic_id: Identifier of the direct messages topic to
                which the message will be sent; required if the message is sent
                to a direct messages chat.
            ephemeral_message_parameters: EphemeralMessageParameters as a
                ``to_dict`` object or dict.
            caption: Voice message caption, 0-1024 characters.
            parse_mode: Parse mode for the caption.
            caption_entities: Special entities for the caption.
            duration: Duration of the voice message in seconds.
            disable_notification: Send silently.
            protect_content: Protect the content from forwarding and saving.
            allow_paid_broadcast: Pass True to ignore broadcasting limits for a
                fee of 0.1 Telegram Stars per message.
            message_effect_id: Unique identifier of the message effect to add.
            suggested_post_parameters: SuggestedPostParameters as a ``to_dict``
                object or dict; for direct messages chats only.
            reply_parameters: Description of the message to reply to, as a
                ``ReplyParameters`` object or a mapping.
            reply_markup: Markup for the message; dict or ``to_dict`` object.

        Returns:
            The sent Message.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#sendvoice
        """
        payload = clean_payload(
            chat_id=chat_id,
            voice=voice,
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            direct_messages_topic_id=direct_messages_topic_id,
            ephemeral_message_parameters=to_wire(ephemeral_message_parameters),
            caption=caption,
            parse_mode=parse_mode,
            caption_entities=[entity.to_dict() for entity in caption_entities]
            if caption_entities is not None
            else None,
            duration=duration,
            disable_notification=disable_notification,
            protect_content=protect_content,
            allow_paid_broadcast=allow_paid_broadcast,
            message_effect_id=message_effect_id,
            suggested_post_parameters=to_wire(suggested_post_parameters),
            reply_parameters=to_wire(reply_parameters),
            reply_markup=to_wire(reply_markup),
        )
        return parse_result(Message, await self.request("sendVoice", payload))

    async def send_video_note(
        self,
        chat_id: int | str,
        video_note: str,
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        direct_messages_topic_id: int | None = None,
        ephemeral_message_parameters: EphemeralMessageParameters | MarkupLike | None = None,
        duration: int | None = None,
        length: int | None = None,
        thumbnail: str | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
        allow_paid_broadcast: bool | None = None,
        message_effect_id: str | None = None,
        suggested_post_parameters: SuggestedPostParameters | MarkupLike | None = None,
        reply_parameters: ReplyParameters | MarkupLike | None = None,
        reply_markup: MarkupLike | None = None,
    ) -> Message:
        """Send a square video message (video note).

        Example:
            >>> msg = await bot.send_video_note(123456, "vn_file_id", length=240)

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            video_note: Video note to send as a ``file_id`` string or HTTP URL.
            business_connection_id: Unique identifier of the business
                connection on behalf of which the message will be sent.
            message_thread_id: Unique identifier for the target message thread.
            direct_messages_topic_id: Identifier of the direct messages topic to
                which the message will be sent; required if the message is sent
                to a direct messages chat.
            ephemeral_message_parameters: EphemeralMessageParameters as a
                ``to_dict`` object or dict.
            duration: Duration of the video in seconds.
            length: Video width and height, i.e. diameter of the video note.
            thumbnail: Thumbnail file_id or URL of the file sent.
            disable_notification: Send silently.
            protect_content: Protect the content from forwarding and saving.
            allow_paid_broadcast: Pass True to ignore broadcasting limits for a
                fee of 0.1 Telegram Stars per message.
            message_effect_id: Unique identifier of the message effect to add.
            suggested_post_parameters: SuggestedPostParameters as a ``to_dict``
                object or dict; for direct messages chats only.
            reply_parameters: Description of the message to reply to, as a
                ``ReplyParameters`` object or a mapping.
            reply_markup: Markup for the message; dict or ``to_dict`` object.

        Returns:
            The sent Message.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#sendvideonote
        """
        payload = clean_payload(
            chat_id=chat_id,
            video_note=video_note,
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            direct_messages_topic_id=direct_messages_topic_id,
            ephemeral_message_parameters=to_wire(ephemeral_message_parameters),
            duration=duration,
            length=length,
            thumbnail=thumbnail,
            disable_notification=disable_notification,
            protect_content=protect_content,
            allow_paid_broadcast=allow_paid_broadcast,
            message_effect_id=message_effect_id,
            suggested_post_parameters=to_wire(suggested_post_parameters),
            reply_parameters=to_wire(reply_parameters),
            reply_markup=to_wire(reply_markup),
        )
        return parse_result(Message, await self.request("sendVideoNote", payload))
